delft/sequenceLabelling/models.py

```diff
@@ -324,14 +327,37 @@ def __init__(self, config, ntags=None):
                            name='casing_embedding')(casing_input)
         casing_embedding = Dropout(config.dropout)(casing_embedding)
 
+        feature_count = 1 if not config.features_indices else len(config.features_indices)
```

Comment:
I think this is where it slightly differs to my implementation and it might explain why the preprocessor did not work as expected.

In my version, the model received a single feature vector (per word token) rather than multiple features. And it is was up to the preprocessor to create that vector.

Here we have knowledge about the selected feature indices. And it seems that we assume that every feature is binary or has the exact same length (`features_vector_size`)? What if they are not binary? e.g. `OPENBRACKET` etc. I was also considering adding POS-tags from a language model. Or are the features passed in here already one-hot encoded? Apologies if I am misreading it.

Reply:
@de-code thank you for your input :-) 

> I think this is where it slightly differs to my implementation and it might explain why the preprocessor did not work as expected.
> 
> In my version, the model received a single feature vector (per word token) rather than multiple features. And it is was up to the preprocessor to create that vector.

> Here we have knowledge about the selected feature indices. And it seems that we assume that every feature is binary or has the exact same length (`features_vector_size`)? What if they are not binary? e.g. `OPENBRACKET` etc. I was also considering adding POS-tags from a language model. Or are the features passed in here already one-hot encoded? Apologies if I am misreading it.

For each token, there is a vector of features that is either calculated dynamically (when the model is trained) or the data is loaded from the model config. In the configuration of the model, I save not only the index of each feature, but also how each feature is encoded.

Features can be binary but also normal values, as long as they have a unique cardinality of a maximum of 12. The one_hot_encoding is done in the data-generator line 134

Reply:
I can look at it more closely in the new year.

I can see now that you are determining the value map for each feature separately.
I am just a bit confused here, `feature_vector_size` seem to default to 12. So you would use a one-hot encoded vector of the length 12 even for a binary feature? For example for features `9-30` that would be a total vector of the size `264`?

Reply:
> I can look at it more closely in the new year.

Thank you :-) 

> I can see now that you are determining the value map for each feature separately.
> I am just a bit confused here, `feature_vector_size` seems to default to 12. So you would use a one-hot encoded vector of the length 12 even for a binary feature? For example for features `9-30` that would be a total vector of the size `264`?

Yes, that's the current idea at the moment, I haven't really thought about it. It's worth discussing, as it might be more efficient to save space.

Reply:
I would probably recommend allowing each feature to take up only the space it needs. The [OneHotEncoder](https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.OneHotEncoder.html) or [DictVectorizer](https://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.DictVectorizer.html) already provide that functionality but you could easily do that yourself having already calculated the cardinality.

The cardinality of 12 should probably only be relevant for the automatic feature selection. But if someone specifies a feature with a higher number via feature indices then that shouldn't be an issue.

What do you think?

Reply:
@de-code I have no opinion about the feature configuration (12 or the space they need)... 

On the other hand, IMHO if a user selects a feature with more than 12 values we should raise an exception I think